blkback: Somewhat clean up the backend-changed handler.

diff -r d1988952cd15 drivers/xen/blkback/xenbus.c
--- a/drivers/xen/blkback/xenbus.c	Thu Jun 24 22:20:18 2010 -0700
+++ b/drivers/xen/blkback/xenbus.c	Thu Jun 24 22:31:15 2010 -0700
@@ -602,52 +602,35 @@
 }
 
 /**
- * Callback received when the hotplug scripts have placed the physical-device
- * node.  Read it and the mode node, and create a vbd.  If the frontend is
- * ready, connect.
+ * Callback received when the hotplug scripts have placed the
+ * physical-device node. If the backend is ready, open the bdev. If
+ * the frontend is ready, connect.
  */
 static void backend_changed(struct xenbus_watch *watch,
 			    const char **vec, unsigned int len)
 {
-	int err;
-	unsigned major;
-	unsigned minor;
 	struct backend_info *be
 		= container_of(watch, struct backend_info, backend_watch);
-	struct xenbus_device *dev = be->dev;
+	struct xenbus_device *xbdev = be->dev;
+	blkif_t *blkif = be->blkif;
+	int err;
 
-	DPRINTK("");
-
-	err = xenbus_scanf(XBT_NIL, dev->nodename, "physical-device", "%x:%x",
-			   &major, &minor);
-	if (XENBUS_EXIST_ERR(err)) {
-		/* Since this watch will fire once immediately after it is
-		   registered, we expect this.  Ignore it, and wait for the
-		   hotplug scripts. */
-		return;
-	}
-	if (err != 2) {
-		xenbus_dev_fatal(dev, err, "reading physical-device");
+	if (!xenbus_exists(XBT_NIL, xbdev->nodename, "physical-device")) {
+		blkif->dev = 0;
 		return;
 	}
 
-	if ((be->major || be->minor) &&
-	    ((be->major != major) || (be->minor != minor))) {
-		printk(KERN_WARNING
-		       "blkback: changing physical device (from %x:%x to "
-		       "%x:%x) not supported.\n", be->major, be->minor,
-		       major, minor);
+	if (xenbus_exists(XBT_NIL, xbdev->nodename, "pause"))
+		return;
+
+	err = blkback_open_bdev(be);
+	if (err) {
+		xenbus_dev_error(xbdev, err, "opening device");
 		return;
 	}
 
-	if (be->major == 0 && be->minor == 0) {
-		err = blkback_open_bdev(be);
-		if (err)
-			return;
-
-		/* We're potentially connected now */
-		update_blkif_status(be->blkif);
-	}
+	/* We're potentially connecting now */
+	update_blkif_status(be->blkif);
 }
 
 
