diff -r c118d9f80b32 include/linux/if_bridge.h
--- a/include/linux/if_bridge.h	Wed Oct 01 16:09:30 2008 +0100
+++ b/include/linux/if_bridge.h	Wed Oct 01 16:09:30 2008 +0100
@@ -108,6 +108,8 @@ extern int (*br_handle_frame_hook)(struc
 extern int (*br_handle_frame_hook)(struct net_bridge_port *p, struct sk_buff **pskb);
 extern int (*br_should_route_hook)(struct sk_buff **pskb);
 
+extern struct net_device *br_locate_physical_device(struct net_device *dev);
+
 #endif
 
 #endif
diff -r c118d9f80b32 net/bridge/br_if.c
--- a/net/bridge/br_if.c	Wed Oct 01 16:09:30 2008 +0100
+++ b/net/bridge/br_if.c	Wed Oct 01 16:09:30 2008 +0100
@@ -15,6 +15,7 @@
 
 #include <linux/kernel.h>
 #include <linux/netdevice.h>
+#include <linux/etherdevice.h>
 #include <linux/ethtool.h>
 #include <linux/if_arp.h>
 #include <linux/module.h>
@@ -276,6 +277,24 @@ static struct net_bridge_port *new_nbp(s
 	.name	= "bridge",
 };
 
+struct net_device *br_locate_physical_device(struct net_device *dev)
+{
+	struct net_bridge *br;
+	const struct net_bridge_port *p;
+
+	if (!(dev->priv_flags & IFF_EBRIDGE))
+		return dev;
+
+	br = netdev_priv(dev);
+
+	list_for_each_entry(p, &br->port_list, list) {
+		if (!compare_ether_addr(dev->dev_addr, p->dev->dev_addr))
+			return p->dev;
+	}
+	return dev;
+}
+EXPORT_SYMBOL(br_locate_physical_device);
+
 int br_add_bridge(struct net *net, const char *name)
 {
 	struct net_device *dev;
