diff -r 9683e34e1bc5 buildconfigs/conf.linux-kdump/kdump
--- a/buildconfigs/conf.linux-kdump/kdump	Thu Dec 09 14:54:11 2010 +0000
+++ b/buildconfigs/conf.linux-kdump/kdump	Fri Dec 17 11:44:00 2010 +0000
@@ -115,3 +115,4 @@
 # CONFIG_X86_CHECK_BIOS_CORRUPTION is not set
 # CONFIG_VMWARE_PVSCSI is not set
 # CONFIG_VMXNET3 is not set
+# CONFIG_DMAR is not set
diff -r 9683e34e1bc5 buildconfigs/linux-defconfig_xen_x86_32
--- a/buildconfigs/linux-defconfig_xen_x86_32	Thu Dec 09 14:54:11 2010 +0000
+++ b/buildconfigs/linux-defconfig_xen_x86_32	Fri Dec 17 11:44:00 2010 +0000
@@ -1,7 +1,7 @@
 #
 # Automatically generated make config: don't edit
 # Linux kernel version: 2.6.32.12-0.7.1
-# Tue Nov  9 13:44:28 2010
+# Fri Dec 17 05:21:29 2010
 #
 # CONFIG_64BIT is not set
 CONFIG_X86_32=y
@@ -196,6 +196,7 @@
 # CONFIG_SPARSE_IRQ is not set
 CONFIG_X86_MPPARSE=y
 CONFIG_X86_XEN=y
+CONFIG_X86_SUPPORTS_MEMORY_FAILURE=y
 CONFIG_SCHED_OMIT_FRAME_POINTER=y
 CONFIG_XEN_SAVE_RESTORE=y
 # CONFIG_M386 is not set
@@ -257,7 +258,9 @@
 # CONFIG_PREEMPT is not set
 CONFIG_X86_LOCAL_APIC=y
 CONFIG_X86_IO_APIC=y
-# CONFIG_X86_MCE is not set
+CONFIG_X86_MCE=y
+CONFIG_X86_MCE_INJECT=m
+CONFIG_X86_XEN_MCE=y
 CONFIG_VM86=y
 # CONFIG_TOSHIBA is not set
 # CONFIG_I8K is not set
@@ -289,6 +292,9 @@
 CONFIG_HAVE_MLOCKED_PAGE_BIT=y
 # CONFIG_KSM is not set
 CONFIG_DEFAULT_MMAP_MIN_ADDR=4096
+CONFIG_ARCH_SUPPORTS_MEMORY_FAILURE=y
+CONFIG_MEMORY_FAILURE=y
+CONFIG_HWPOISON_INJECT=y
 # CONFIG_PRECACHE is not set
 # CONFIG_PRESWAP is not set
 # CONFIG_HIGHPTE is not set
@@ -366,7 +372,6 @@
 # CONFIG_PCIEPORTBUS is not set
 CONFIG_ARCH_SUPPORTS_MSI=y
 CONFIG_PCI_MSI=y
-# CONFIG_DMAR is not set
 CONFIG_PCI_LEGACY=y
 # CONFIG_PCI_DEBUG is not set
 CONFIG_PCI_GUESTDEV=y
