diff -r 3512b39f56da drivers/xen/netback/netback.c
--- a/drivers/xen/netback/netback.c	Fri May 07 15:55:30 2010 +0100
+++ b/drivers/xen/netback/netback.c	Fri May 07 16:09:16 2010 +0100
@@ -241,11 +241,7 @@
 		len -= copy;
 	}
 
-#ifdef NET_SKBUFF_DATA_USES_OFFSET
-	offset = 0;
-#else
 	offset = nskb->data - skb->data;
-#endif
 
 	nskb->transport_header = skb->transport_header + offset;
 	nskb->network_header   = skb->network_header   + offset;
