# HG changeset patch
# Parent 3c0070e1b1a78e4ef583f65e02b04ddcd04e3d1c

diff -r 3c0070e1b1a7 arch/x86/kernel/Makefile
--- a/arch/x86/kernel/Makefile	Tue Jul 31 16:00:46 2012 +0100
+++ b/arch/x86/kernel/Makefile	Tue Jul 31 18:49:20 2012 +0100
@@ -15,6 +15,9 @@ CFLAGS_REMOVE_pvclock.o = -pg
 CFLAGS_REMOVE_kvmclock.o = -pg
 CFLAGS_REMOVE_ftrace.o = -pg
 CFLAGS_REMOVE_early_printk.o = -pg
+ifdef CONFIG_XEN
+CFLAGS_REMOVE_time.o = -pg
+endif
 endif
 
 #
diff -r 3c0070e1b1a7 drivers/xen/core/evtchn.c
--- a/drivers/xen/core/evtchn.c	Tue Jul 31 16:00:46 2012 +0100
+++ b/drivers/xen/core/evtchn.c	Tue Jul 31 18:49:20 2012 +0100
@@ -305,7 +305,7 @@ void __init init_IRQ(void)
  * callback mask. We do this in a very simple manner, by making a call
  * down into Xen. The pending flag will be checked by Xen on return.
  */
-void force_evtchn_callback(void)
+notrace void force_evtchn_callback(void)
 {
 	VOID(HYPERVISOR_xen_version(0, NULL));
 }
